### PR TITLE
core: support shorthand hex colors (thx @tobes)

### LINKED ISSUE #1200: support shorthand hexadecimal
Right now, the config support six-digit hexadecimal.
```python
# working - success
module {
    color_good = '#00FF00'
    color_bad = '#FF0000'
    color_degraded = '#FFFF00'
    color_charging = '#00FFFF'
}
```
Request feature: The config is to support shorthand three-digit hexadecimal too.
```python
# not working - failure
module {
    color_good = '#0F0'
    color_bad = '#F00'
    color_degraded = '#FF0'
    color_charging = '#0FF'
}
```
Optional: The config is to support x11 color names too.
```python
# not working - future rgb names
module {
    color_good = 'steelblue`
    color_bad = 'rosered'
    color_degraded = 'cornyellow`
    color_charging = 'tobespink`
}
```
Too much! The config is to support rgb numbers too.
```python
# not working - future rgb numbers
module {
    color_good = rgb(0, 255, 0)
    color_bad = rgb(255, 0, 0)
    color_degraded = rgb(255, 255, 0)
    color_charging = rgb(255, 0, 255)
}
```

### PR DESCRIPTION
**Pull request by lasers**
Hi. I'm not familiar with the module/py3 code. This closes #1200 for me.
```python


        # This code block intentionally left blank.

```
This works okay with `None` and we do this for `color_error` too. I skip non-issue composite.

### CODE REVIEW

File: py3status/module.py [hunk 1]
```diff
@@ -182,6 +182,8 @@ def error_output(self, message, method_affected=None):
             color = color_fn(self.module_full_name, 'color_bad')
         if hasattr(color, 'none_setting'):
             color = None
+        else:
+            color = Py3()._get_color(color)
```
Comment (tobes) [inline py3status/module.py]: This is very wrong.  We do not want to create an instance of Py3 each time we get here.  The module already has an instance of it's own if we need it but I don't think we do.

Do we know where we hit this issue?  If it is just in the config file then we can fix that when we import it.  If there are other places that we are affected then we may want to take a different approach

File: py3status/py3.py [hunk 2]
```diff
@@ -149,6 +149,8 @@ def _get_config_setting(self, name, default=None):
                     param = None
                 elif param is None:
                     param = self._none_color
+                else:
+                    param = self._get_color(param)
```
Comment (tobes) [inline py3status/py3.py]: This would need an explanatory comment

File: py3status/module.py (hunk 1 above)
Comment (lasers) [reply py3status/module.py]: Use this to slap `dpms` around.
```
diff --git a/py3status/modules/dpms.py b/py3status/modules/dpms.py
index 27345f17..8cc54dd2 100644
--- a/py3status/modules/dpms.py
+++ b/py3status/modules/dpms.py
@@ -61,6 +61,7 @@ class Py3status:
     def post_config_hook(self):
         self.color_on = self.py3.COLOR_ON or self.py3.COLOR_GOOD
         self.color_off = self.py3.COLOR_OFF or self.py3.COLOR_BAD
+        self.index = 0
 
     def dpms(self):
         """
@@ -73,6 +74,11 @@ class Py3status:
             _format = self.icon_off
             color = self.color_off
 
+        if self.index > 3:
+            self.py3.error('Oh no! That sucks.')
+
+        self.index += 1
+
         icon = self.py3.safe_format(_format)
 
         return {
```
I can't figure out what to use with...
```
from py3status.py3 import Py3, PY3_CACHE_FOREVER, ModuleErrorException
```
I see `Formatter()` does the same thing. Wrong too?

**Comment by tobes** [comment]
A real fix is nearer to this.  I'm not totally happy with it though.

```diff
diff --git a/py3status/core.py b/py3status/core.py
index 6a81d547..34c508bd 100644
--- a/py3status/core.py
+++ b/py3status/core.py
@@ -65,6 +65,9 @@ class NoneSetting:
     # this attribute is used to identify that this is a none setting
     none_setting = True
 
+    def __len__(self):
+        return 0
+
     def __repr__(self):
         # this is for output via module_test
         return 'None'
@@ -106,6 +109,14 @@ class Common:
         if hasattr(param, 'none_setting'):
             # check py3status general section
             param = config['general'].get(attribute, self.none_setting)
+        # If this is a color and it is like #123 convert it to #112233
+        if param and (attribute == 'color' or attribute.startswith('color_')):
+            if len(param) == 4 and param[0] == '#':
+                param = (
+                    '#' + param[1] + param[1] + param[2] +
+                    param[2] + param[3] + param[3]
+                )
+
         return param
 
     def report_exception(self, msg, notify_user=True, level='error',
```

**Comment by lasers** [comment]
@tobes Thx. Any reason why we shouldn't use `color` in attribute here?

**Comment by tobes** [comment]
> Any reason why we shouldn't use `'color' in attribute` here?

Yes, because that could catch things outside of the strict criteria.  tbh it would probably be fine but I'd prefer us to be super strict with things.

**Comment by ultrabug** [comment]
Assigning to you @tobes, merge when rdy thx!

**Comment by tobes** [comment]
Let's get things closed :)

**Pull request merged**